step2_spark/recom.py

```diff
@@ -0,0 +1,102 @@
+import pyspark.sql.functions as sql_func
+from pyspark.sql.types import *
+from pyspark.mllib.recommendation import MatrixFactorizationModel, Rating #, ALS
+from pyspark.mllib.evaluation import RegressionMetrics, RankingMetrics
+from pyspark.ml.evaluation import ClusteringEvaluator
+from pyspark.sql import Row
+import pandas.io.sql as sqlio
+import psycopg2
+
+from pyspark.ml.recommendation import ALS, ALSModel
+from pyspark.context import SparkContext, SparkConf
+from pyspark.sql.session import SparkSession
+from pyspark.ml.evaluation import RegressionEvaluator
+from pyspark.ml.clustering import KMeans
+import pandas as pd
+import numpy as np
+import boto3
+from io import StringIO
+
+def _map_to_pandas(rdds):
+    """ Needs to be here due to pickling issues """
+    return [pd.DataFrame(list(rdds))]
+
+def to_Pandas(df, n_partitions=None):
+    """
+    Returns the contents of `df` as a local `pandas.DataFrame` in a speedy fashion. The DataFrame is
+    repartitioned if `n_partitions` is passed.
+    :param df:              pyspark.sql.DataFrame
+    :param n_partitions:    int or None
+    :return:                pandas.DataFrame
+    """
+    if n_partitions is not None: df = df.repartition(n_partitions)
```

Comment:
This can blow up your local memory if the df being collected is even a moderate size.